addDropIndexDesign.md

```diff
@@ -0,0 +1,68 @@
+
+# Add/Drop Index
+
+## Overview
+
+>*What motivates this to be implemented? What will this component achieve?* 
+
+Currently, the system just creates the index without populating it. This component supports actually adding the index (including populating the index) and/or dropping in a transactionally consistent manner. This is accomplished by blocking any transaction wanting to write to the table as the component builds the index through a sequential scan. Future work includes supporting building with concurrent modifications.
+
+## Scope
+>*Which parts of the system will this feature rely on or modify? Write down specifics so people involved can review the design doc*
+
+This mainly touches the traffic cop (to control blocking), index_builder (to build the index), the database catalog (to add the new flags to the indexes, such as live), and transaction stuff (to manage held table locks by a transaction)
+
+The parts of the system this feature will modify include the following files:
+ - index_builder.\*
+ - ddl_executors.\*
+ - traffic_cop.\*
+ - transaction_context.\*
+ - transaction_manager.\*
+ - catalog_accessor.\*
+ - database_catalog.\*
+
+
+## Architectural Design
+>*Explain the input and output of the component, describe interactions and breakdown the smaller components if any. Include diagrams if appropriate.*
+
+In broad strokes, this component works by first waiting until all uncommitted transactions that have already modified the table either commit or abort. During this time and until index creation is completed, any transactions that have not already modified the table and wish to do so are blocked and must wait until index creation is complete. After the transactions all complete, index creation proceeds by scanning through the table and inserting all visible versions of each tuple into the index. Once index creation is complete, waiting transactions are allowed to proceed.
+
```

Comment:
> Is it possible to not blocking the transactions after index creation started

It is not possible to do no blocking of transactions after index creation has started as you run into too many concurrency issues that way. It is however possible to only block them while running through a smaller delta store (rather than while running through the whole table). This is what CREATE INDEX CONCURRENTLY should do (see §Future Work).

> We could put a flag to suggest if one table is being built index on. Modification by those txns are not visible to the index creation txn during scan, and the modifying txns can defer/stash the index update of the modified tuples and let those run before the index become visible.

This is similar to what our original proposal did. There's two classes of issues here:
  - Transactions that have started running but have not yet made modifications to the table at time of index creation:
           These transactions wont have access to knowledge of the index being built unless we provide a lot more information to the codegen'ed code (which would require massive infrastructure changes). This is still solvable though.
  - Transactions that have started running and have already made modifications to the table before index creations:
            These transactions have no way of finding out the index is being built. In order to solve the problem of these transactions committing (for delete) or aborting (for insert), you would have to be able to go from transaction timestamp to transaction pointer. See the first paragraph of §Trade-offs and Potential Problems